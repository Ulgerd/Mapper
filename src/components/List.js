import React from 'react';
import { connect } from 'react-redux';
import ListStr from './ListStr.js';
import { removeDot, rearrangeDots } from '../actions/rootActions.js'
import { DragDropContext, Droppable } from 'react-beautiful-dnd'
import styled from 'styled-components'

const StyledUl = styled.ul`
   list-style-type: none;
   display: flex;
   flex-direction: column;
   justify-content: flex-start;
   margin-left: 0;
   padding-left: 0;
`

const reorder = (list, startIndex, endIndex) => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);
  result.splice(endIndex, 0, removed);
  return result;
};

function List(props) {
  const onDragEnd =(result) => {
    // dropped outside the list
    if (!result.destination) {
      return;
    }

    const items = reorder(
      props.dots,
      result.source.index,
      result.destination.index
    );

    props.rearrangeDots(items);
  }

  return (
    <DragDropContext onDragEnd={onDragEnd}>
        <Droppable droppableId="droppable">
        {(provided, snapshot) => (
         <StyledUl
           {...provided.droppableProps}
           ref={provided.innerRef}
         >
            {props.dots.map((obj, index) => {
              return <ListStr
                        key={obj.id}
                        id={obj.id}
                        index={index}
                        name={obj.name}
                        onDotRemove={props.removeDot}
                      />
            })}
          </StyledUl>
        )}
        </Droppable>
      </DragDropContext>

  );
}

const mapStateToProps = store => {
  return {
    dots: store.dots,
  }
}

const mapDispatchToProps = dispatch => ({
  removeDot: (dotPos) => {dispatch(removeDot(dotPos))},
  rearrangeDots: (dotsArr) => {dispatch(rearrangeDots(dotsArr))}
})

export default connect(mapStateToProps, mapDispatchToProps)(List);
