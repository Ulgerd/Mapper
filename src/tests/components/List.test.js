import React from 'react';
import { shallow } from 'enzyme'
import { List } from '../../components/List.js';
import renderer from 'react-test-renderer';
jest.mock('../../components/ListStr', () => 'ListStr')

jest.mock('react-beautiful-dnd', () => ({
  Droppable: ({ children }) => children({
    draggableProps: {
      style: {},
    },
    innerRef: jest.fn(),
  }, {}),
  Draggable: ({ children }) => children({
    draggableProps: {
      style: {},
    },
    innerRef: jest.fn(),
  }, {}),
  DragDropContext: ({ children }) => children,
}));

describe('List component tests',()=>{
    let wrapper

    const props = {
      dots: [{id: 123, name: 'Moscow', coords: [14.2, 44.1]},
             {id: 321, name: 'Istanbul', coords: [42.2, 14.1]}]
    };

    beforeEach(()=>{
        wrapper = shallow(<List {...props} />)
    })

    it('render the DUMB component', () => {
       expect(wrapper.length).toEqual(1)
    });

    // it('contains ul', () => {
    //   expect(wrapper.find('ul')).toHaveLength(1)
    // });
    //
    // it('contains ListStr', () => {
    //   expect(wrapper.find('ListStr')).toHaveLength(2)
    // });

    it('renders correctly', () => {
      const tree = renderer
        .create(<List {...props} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
});
