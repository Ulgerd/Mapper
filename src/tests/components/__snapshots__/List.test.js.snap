// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component tests renders correctly 1`] = `
<ul
  className="sc-bdVaJa itWsyt"
>
  <ListStr
    id={123}
    index={0}
    name="Moscow"
  />
  <ListStr
    id={321}
    index={1}
    name="Istanbul"
  />
</ul>
`;
